# 1. Preprocessing

This section contains the following: 

    1.1. Overview of the dataset
    1.2. Addressing missing values
    1.3. Filtering required columns
    1.4. Adding required columns
    1.5. Downloading preprocessing dataset as a CSV file
    1.6. Insights

We start by importing the required packages

In [2]:
import pandas as pd
import numpy as np

Next, the downloaded dataset *Analysis* is read into a Pandas DataFrame *preprocessing_dataset*. 

In [3]:
preprocessing_dataset = pd.read_csv("C:\\Users\\Sharmila\\Desktop\\time-series-19-covid-combined.csv")

In [4]:
print(preprocessing_dataset)

             Date Country/Region Province/State   Lat  Long  Confirmed  \
0      2020-01-22    Afghanistan            NaN  33.0  65.0        0.0   
1      2020-01-23    Afghanistan            NaN  33.0  65.0        0.0   
2      2020-01-24    Afghanistan            NaN  33.0  65.0        0.0   
3      2020-01-25    Afghanistan            NaN  33.0  65.0        0.0   
4      2020-01-26    Afghanistan            NaN  33.0  65.0        0.0   
...           ...            ...            ...   ...   ...        ...   
23491  2020-04-15       Zimbabwe            NaN -20.0  30.0       23.0   
23492  2020-04-16       Zimbabwe            NaN -20.0  30.0       23.0   
23493  2020-04-17       Zimbabwe            NaN -20.0  30.0       24.0   
23494  2020-04-18       Zimbabwe            NaN -20.0  30.0       25.0   
23495  2020-04-19       Zimbabwe            NaN -20.0  30.0       25.0   

       Recovered  Deaths  
0            0.0     0.0  
1            0.0     0.0  
2            0.0     0.0  
3  

## <font color=black>1.1. Overview of the Dataset</font>

The columns in the dataset, their types, index values, first and last five rows is displayed. The describe() function is used to provide a quick statistical overview of the data.

In [5]:
# The names of the columns

columns_dataset = preprocessing_dataset.columns
print(columns_dataset)
print(len(columns_dataset))

Index(['Date', 'Country/Region', 'Province/State', 'Lat', 'Long', 'Confirmed',
       'Recovered', 'Deaths'],
      dtype='object')
8


In [41]:
# The data type of each column

print(preprocessing_dataset.dtypes)

Year             int64
Mo             float64
Dy             float64
Hr             float64
Mn             float64
Sec            float64
Tsu             object
Vol             object
Addl            object
Name            object
Latitude       float64
Longitude      float64
Focal          float64
Mag            float64
MMI Int        float64
Num            float64
De             float64
Num.1          float64
De.1           float64
$Mill          float64
De.2           float64
Num.2          float64
De.3           float64
Num.3          float64
De.4           float64
Unnamed: 25    float64
dtype: object


In [6]:
# Next, the indexes associated with the dataframe are checked

preprocessing_dataset.index.values

array([    0,     1,     2, ..., 23493, 23494, 23495], dtype=int64)

The top five and bottom five rows are displayed below. It is seen that each observation forms a row, is defined by an index, and attributes of those observations are displayed in the columns of the DataFrame. 

In [7]:
# First five rows

print(preprocessing_dataset.head())

         Date Country/Region Province/State   Lat  Long  Confirmed  Recovered  \
0  2020-01-22    Afghanistan            NaN  33.0  65.0        0.0        0.0   
1  2020-01-23    Afghanistan            NaN  33.0  65.0        0.0        0.0   
2  2020-01-24    Afghanistan            NaN  33.0  65.0        0.0        0.0   
3  2020-01-25    Afghanistan            NaN  33.0  65.0        0.0        0.0   
4  2020-01-26    Afghanistan            NaN  33.0  65.0        0.0        0.0   

   Deaths  
0     0.0  
1     0.0  
2     0.0  
3     0.0  
4     0.0  


In [8]:
# Last five rows

print(preprocessing_dataset.tail())

             Date Country/Region Province/State   Lat  Long  Confirmed  \
23491  2020-04-15       Zimbabwe            NaN -20.0  30.0       23.0   
23492  2020-04-16       Zimbabwe            NaN -20.0  30.0       23.0   
23493  2020-04-17       Zimbabwe            NaN -20.0  30.0       24.0   
23494  2020-04-18       Zimbabwe            NaN -20.0  30.0       25.0   
23495  2020-04-19       Zimbabwe            NaN -20.0  30.0       25.0   

       Recovered  Deaths  
23491        1.0     3.0  
23492        1.0     3.0  
23493        2.0     3.0  
23494        2.0     3.0  
23495        2.0     3.0  


In [45]:
# The basic statistical overview of the dataset is provided using Pandas describe()

print(preprocessing_dataset.describe(include = 'all'))

               Year           Mo           Dy           Hr          Mn  \
count   6144.000000  5739.000000  5587.000000  4113.000000  3908.00000   
unique          NaN          NaN          NaN          NaN         NaN   
top             NaN          NaN          NaN          NaN         NaN   
freq            NaN          NaN          NaN          NaN         NaN   
mean    1804.691243     6.505663    15.721675    11.304644    28.86131   
std      376.404999     3.449445     8.749653     7.032811    17.14624   
min    -2150.000000     1.000000     1.000000     0.000000     0.00000   
25%     1820.000000     4.000000     8.000000     5.000000    14.75000   
50%     1928.000000     7.000000    16.000000    11.000000    30.00000   
75%     1988.000000     9.000000    23.000000    17.000000    44.00000   
max     2019.000000    12.000000    31.000000    23.000000    59.00000   

                Sec   Tsu  Vol  Addl                    Name  ...  \
count   2791.000000  1853   64  6144      

## 1.2. Addressing Missing Values

The missing values in *Name, Country/Region, Province/State, Latitude, Longitude, Confirmed, Recovered, Deaths are explored in detail. A question mark(?) is used to replace the missing values in other variables.  

In [9]:
# The null values in the dataset is viewed: 

print(preprocessing_dataset.isna().sum())

Date                  0
Country/Region        0
Province/State    16198
Lat                   0
Long                  0
Confirmed            89
Recovered          1246
Deaths               89
dtype: int64


### 1.2.1. Province/State

There exists 16198 missing entries in *Province/State*. The rows containing missing values is retrieved. Based on the latitude and longitude of mentioned in that row, the entire dataset is searched to find a similar entry.

In [12]:
# Accessing index of the missing entries in the Name column

np.where(pd.isnull(preprocessing_dataset['Province/State']))

(array([    0,     1,     2, ..., 23493, 23494, 23495], dtype=int64),)

In [19]:
preprocessing_dataset = preprocessing_dataset.fillna('?')
print(preprocessing_dataset)

             Date Country/Region Province/State   Lat  Long Confirmed  \
0      2020-01-22    Afghanistan              ?  33.0  65.0         0   
1      2020-01-23    Afghanistan              ?  33.0  65.0         0   
2      2020-01-24    Afghanistan              ?  33.0  65.0         0   
3      2020-01-25    Afghanistan              ?  33.0  65.0         0   
4      2020-01-26    Afghanistan              ?  33.0  65.0         0   
...           ...            ...            ...   ...   ...       ...   
23491  2020-04-15       Zimbabwe              ? -20.0  30.0        23   
23492  2020-04-16       Zimbabwe              ? -20.0  30.0        23   
23493  2020-04-17       Zimbabwe              ? -20.0  30.0        24   
23494  2020-04-18       Zimbabwe              ? -20.0  30.0        25   
23495  2020-04-19       Zimbabwe              ? -20.0  30.0        25   

      Recovered Deaths  
0             0      0  
1             0      0  
2             0      0  
3             0      0 

In [20]:
# Verifying if the null value still exists

print(preprocessing_dataset['Province/State'].isna().sum())

0


### 1.2.2. Confirmed, Recovered and Deaths

In [22]:
# Calculating the total number of missing values for Column Confirmed

print(preprocessing_dataset.Confirmed.isna().sum())

0


In [24]:
# Calculating the total number of missing Recovered

print(preprocessing_dataset.Recovered.isna().sum())

0


In [25]:
# Calculating the total number of missing Deaths

print(preprocessing_dataset.Deaths.isna().sum())

0


The reasons behind the missing values is not yet understood.The COVID-19 is an ongoing, extraordinary situation, and the numbers are being updated regularly.

### 1.2.3. Other variables

Not enough is known about the missing values in other variables. These empty entries are replaced with a question mark(?) for the time being.

In [26]:
# Calculating the number of missing values in the dataset

preprocessing_dataset.isna().sum()

Date              0
Country/Region    0
Province/State    0
Lat               0
Long              0
Confirmed         0
Recovered         0
Deaths            0
dtype: int64

In [27]:
# Replacing the missing values with a question mark(?)

preprocessing_dataset = preprocessing_dataset.fillna('?')
print(preprocessing_dataset)

             Date Country/Region Province/State   Lat  Long Confirmed  \
0      2020-01-22    Afghanistan              ?  33.0  65.0         0   
1      2020-01-23    Afghanistan              ?  33.0  65.0         0   
2      2020-01-24    Afghanistan              ?  33.0  65.0         0   
3      2020-01-25    Afghanistan              ?  33.0  65.0         0   
4      2020-01-26    Afghanistan              ?  33.0  65.0         0   
...           ...            ...            ...   ...   ...       ...   
23491  2020-04-15       Zimbabwe              ? -20.0  30.0        23   
23492  2020-04-16       Zimbabwe              ? -20.0  30.0        23   
23493  2020-04-17       Zimbabwe              ? -20.0  30.0        24   
23494  2020-04-18       Zimbabwe              ? -20.0  30.0        25   
23495  2020-04-19       Zimbabwe              ? -20.0  30.0        25   

      Recovered Deaths  
0             0      0  
1             0      0  
2             0      0  
3             0      0 

In [28]:
# Verifying that there are no missing values present in the entire dataset

print(preprocessing_dataset.isna().sum())

Date              0
Country/Region    0
Province/State    0
Lat               0
Long              0
Confirmed         0
Recovered         0
Deaths            0
dtype: int64


## 1.3. Filtering required columns

Out of the 08 columns, Country, Latitude, Longitude, Confirmed, Recovered, and Deaths is thought to be not relevant for most of the analysis. They are referred to later when comparing the events. For the rest of the analysis, only the following columns are filtered: 

Country, Province, Latitude, Longitude, Confirmed, Recovered, and Deaths. The code for filtering these columns is displayed below:

In [35]:
dataset_for_analysis = preprocessing_dataset.filter(['Country/Region', 'Province/State', 'Lat', 'Long', 'Confirmed', 'Recovered', 'Deaths'])
print(dataset_for_analysis)

      Country/Region Province/State   Lat  Long Confirmed Recovered Deaths
0        Afghanistan              ?  33.0  65.0         0         0      0
1        Afghanistan              ?  33.0  65.0         0         0      0
2        Afghanistan              ?  33.0  65.0         0         0      0
3        Afghanistan              ?  33.0  65.0         0         0      0
4        Afghanistan              ?  33.0  65.0         0         0      0
...              ...            ...   ...   ...       ...       ...    ...
23491       Zimbabwe              ? -20.0  30.0        23         1      3
23492       Zimbabwe              ? -20.0  30.0        23         1      3
23493       Zimbabwe              ? -20.0  30.0        24         2      3
23494       Zimbabwe              ? -20.0  30.0        25         2      3
23495       Zimbabwe              ? -20.0  30.0        25         2      3

[23496 rows x 7 columns]


### Description of the columns

Date: Date on which a case was reported
Country/Region: Name of the country/region
Province/State: Name of the Province/State
Lat: Latitude
Long: Longitude
Confirmed: Positive cases
Recovered: Number of cases where people who have tested positive have recovered
Deaths: Total number of deaths


## 1.5. Downloading preprocessed dataset as a csv file

The following code downloads the preprocessed dataset as a csv file onto the local system. The downloaded file is then used for further analysis from here on.

In [36]:
dataset_for_analysis.to_csv('Dataset For Analysis.csv')

## 1.6. Insights

### Overview of the dataset

    1. The dataset consists of 23496 rows and 08 columns.

### Addressing missing values

    1. There exist 16198 missing values in Province/State column, 89 in Confirmed, 1245 in Recovered, and 98 in Deaths. 
    2. The database was updated accordingly and verified to make sure the null value was updated.
    3. The reasons behind the missing values in other variables is not well understood and is replaced with a '?'.
